dockers/texcmp/texcmp.js

```diff
@@ -0,0 +1,244 @@
+        return execFile("convert", [
+            "-density", dpi, "-units", "PixelsPerInch", "-flatten",
+            pdfFile, pngFile
+        ]);
+    }).then(function() {
+        console.log("Rasterized " + key);
+        // Step 4: apply FFT to that
+        return fftPNG(pngFile);
+    });
+    // Step 5: apply FFT to reference image as well
+    var fftBrowser = fftPNG(browserFile);
+
+    return Q.all([fftBrowser, fftLatex]).spread(function(browser, latex) {
+        // Now we have the FFT result from both 
+        // Step 6: find alignment which maximizes overlap.
+        // This uses a FFT-based correlation comoputation.
+        var x, y;
+        var real = newMatrix();
+        var imag = newMatrix();
+
+        // Step 6a: (real + i*imag) = latex * conjugate(browser)
+        for (y = 0; y < alignHeight; ++y) {
+            for (x = 0; x < alignWidth; ++x) {
+                var br = browser.real.get(y, x);
+                var bi = browser.imag.get(y, x);
+                var lr = latex.real.get(y, x);
+                var li = latex.imag.get(y, x);
+                real.set(y, x, br * lr + bi * li);
+                imag.set(y, x, br * li - bi * lr);
+            }
+        }
+
+        // Step 6b: (real + i*imag) = inverseFFT(real + i*imag)
+        fft(-1, real, imag);
+
+        // Step 6c: find position where the (squared) absolute value is maximal
+        var ox = 0;
+        var oy = 0;
+        var ov = -1;
+        for (y = 0; y < alignHeight; ++y) {
+            for (x = 0; x < alignWidth; ++x) {
+                var or = real.get(y, x);
+                var oi = imag.get(y, x);
+                var sq = or * or + oi * oi;
+                if (ov < sq) {
+                    ov = sq;
+                    ox = x;
+                    oy = y;
+                }
+            }
+        }
+
+        // Step 6d: Treat negative offsets in a non-cyclic way
+        if (oy > (alignHeight >>> 1)) {
+            oy -= alignHeight;
+        }
+        if (ox > (alignWidth >>> 1)) {
+            ox -= alignWidth;
+        }
+        console.log("Positioned " + key + ": " + ox + ", " + oy);
+
+        // Step 7: use these offsets to compute difference illustration
+        var uw = Math.max(browser.width, latex.width) + Math.abs(ox);
+        var uh = Math.max(browser.height, latex.height) + Math.abs(oy);
+        var bx = Math.max(ox, 0);
+        var by = Math.max(oy, 0);
+        var lx = Math.max(-ox, 0);
+        var ly = Math.max(-oy, 0);
+        return execFile("convert", [
+            "(", pngFile, "-grayscale", "Rec709Luminance",
+            "-extent", uw + "x" + uh + "-" + lx + "-" + ly,
+            ")",
+            "(", browserFile, "-grayscale", "Rec709Luminance",
+            "-extent", uw + "x" + uh + "-" + bx + "-" + by,
+            ")",
+            "(", "-clone", "0-1", "-compose", "darken", "-composite", ")",
+            "-channel", "RGB", "-combine", "-trim", diffFile
```

Comment:
The parenthesized parts describe a sequence of operations which is applied to some input and the result is then used as the input on the outer scope. So the first parenthesis reads the LaTeX file, converts that to grayscale and pads it to the union. The second does the same for the browser screenshot. The third takes the first two and takes the minimal gray value for each pixel. These three then form the red, green and blue channel of the final image. See http://www.imagemagick.org/script/command-line-processing.php#stack for details.